substrate-archive-backend/src/read_only_backend/blockchain_backend.rs

```diff
@@ -33,22 +34,59 @@ use sp_runtime::{
 use crate::{
 	database::ReadOnlyDb,
 	read_only_backend::ReadOnlyBackend,
-	util::{self, columns},
+	util::{self, columns, read_db},
 };
 
 type ChainResult<T> = Result<T, BlockchainError>;
 
+#[derive(Debug, Encode, Decode)]
+struct ExtrinsicHeader {
+	hash: sp_core::H256,
+	data: Vec<u8>,
+}
+
 impl<Block: BlockT, D: ReadOnlyDb> BlockchainBackend<Block> for ReadOnlyBackend<Block, D> {
 	fn body(&self, id: BlockId<Block>) -> ChainResult<Option<Vec<<Block as BlockT>::Extrinsic>>> {
-		let res = util::read_db::<Block, D>(&*self.db, columns::KEY_LOOKUP, columns::BODY, id)
-			.map_err(|e| BlockchainError::Backend(e.to_string()))?;
-
-		match res {
-			Some(body) => match Decode::decode(&mut &body[..]) {
+		let body = match read_db(&*self.db, columns::KEY_LOOKUP, columns::BODY, id)
+			.map_err(|e| BlockchainError::Backend(e.to_string()))?
+		{
+			Some(body) => body,
+			None => return Ok(None),
+		};
+		match self.storage_mode {
+			TransactionStorageMode::BlockBody => match Decode::decode(&mut &body[..]) {
 				Ok(body) => Ok(Some(body)),
-				Err(_) => Err(BlockchainError::Backend("Could not decode extrinsics".into())),
+				Err(err) => return Err(BlockchainError::Backend(format!("Error decoding body: {}", err))),
+			},
+			TransactionStorageMode::StorageChain => match Vec::<ExtrinsicHeader>::decode(&mut &body[..]) {
```

Comment:
I guess you could have a helper fn here, `fn decode_storage_chain_item<Block: BlockT, D: ReadOnlyDb>(body: &Vec<u8>, db: &D) -> ChainResult<Option<Vec<<Block as BlockT>::Extrinsic>>>` and sweep this under the rug with a `TransactionStorageMode::StorageChain => decode_storage_chain_item::<Block, D>(&body, &*self.db),`

Not convinced it's any better though. :/ Maybe it's better to keep it as closely aligned to substrate as possible if we need to copy&paste this.

Better still would be to use `sc_client_api::blockchain::Backend` directly, what was the problem there?

Reply:
I think you're talking about `sc_client_db::Backend`?, in which case we could technically use that, but it wouldn't be very useful for archive so I made a design decision early on to form a custom Backend for Archive.

By 'not very useful' I mean that Archive uses some custom iteration logic to speed up gathering of it's blocks. For instance from our backend implementation we access the `iter` method on the database [here](https://github.com/paritytech/substrate-archive/blob/0df95bd6c5a661ab991d2e4900baff2276892877/substrate-archive-backend/src/read_only_backend.rs#L128). 

We could use Substrate's backend and compose it with our own backend, then re-implement the `Backend` traits on it to carry over it's functionality, but when I was first coding this stuff up that seemed like it was too complicated, and we would only use substrate Backend for backwards compatibility. There's also a bunch of references/stuff on substrate's backend that we wouldn't use pretty much at all in archive